File: src/main/java/plugin/javafxtools/service/HttpTemplateStore.java
package plugin.javafxtools.service;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import plugin.javafxtools.model.HttpTemplate;
import plugin.javafxtools.util.AppDataPaths;
import plugin.javafxtools.util.AtomicFileWriter;
import plugin.javafxtools.util.HttpUrlSupport;

import java.io.IOException;
import java.io.Reader;
import java.lang.reflect.Type;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

/**
 * HTTP 请求模板的本地 JSON 存储。
 *
 * @author wwj
 */
public class HttpTemplateStore {
    private static final Set<String> SUPPORTED_METHODS = Set.of(
            "GET", "POST", "PUT", "DELETE", "PATCH", "HEAD", "OPTIONS");
    private static final Path DEFAULT_TEMPLATE_FILE = AppDataPaths.dataFile("http_templates.json");
    private static final Type TEMPLATE_MAP_TYPE = new TypeToken<Map<String, HttpTemplate>>() {
    }.getType();

    private final Gson gson = new Gson();
    private final Path templateFile;

    public HttpTemplateStore() {
        this(DEFAULT_TEMPLATE_FILE);
    }

    HttpTemplateStore(Path templateFile) {
        this.templateFile = Objects.requireNonNull(templateFile, "templateFile");
    }

    /**
     * 读取请求模板。
     *
     * @return 以模板名为键的模板集合
     */
    public Map<String, HttpTemplate> loadTemplates() {
        if (!Files.exists(templateFile)) {
            return new HashMap<>();
        }

        try {
            return new HashMap<>(readStoredTemplates());
        } catch (Exception e) {
            throw new IllegalStateException("加载模板失败: " + e.getMessage(), e);
        }
    }

    /**
     * 保存请求模板。
     *
     * @param templates 模板集合
     */
    public void saveTemplates(Map<String, HttpTemplate> templates) {
        try {
            validateCurrentSchema(templates);
            validateExistingFileBeforeOverwrite();
            AtomicFileWriter.writeUtf8(templateFile,
                    gson.toJson(templates));
        } catch (Exception e) {
            throw new IllegalStateException("保存模板失败: " + e.getMessage(), e);
        }
    }

    private Map<String, HttpTemplate> readStoredTemplates() throws IOException {
        try (Reader reader = Files.newBufferedReader(templateFile, StandardCharsets.UTF_8)) {
            Map<String, HttpTemplate> templates = gson.fromJson(reader, TEMPLATE_MAP_TYPE);
            validateCurrentSchema(templates);
            return templates;
        }
    }

    private void validateExistingFileBeforeOverwrite() throws IOException {
        if (!Files.exists(templateFile)) {
            return;
        }
        try {
            readStoredTemplates();
        } catch (Exception e) {
            throw new IOException("现有 HTTP 模板配置无效，拒绝覆盖", e);
        }
    }

    private void validateCurrentSchema(Map<String, HttpTemplate> templates) {
        if (templates == null) {
            throw new IllegalArgumentException("配置根节点必须是对象");
        }

        templates.forEach((name, template) -> {
            if (name == null
                    || name.isBlank()
                    || !name.equals(name.trim())
                    || template == null
                    || !HttpUrlSupport.isValid(template.url)
                    || !SUPPORTED_METHODS.contains(template.method)
                    || template.params == null
                    || template.headers == null
                    || !isPositiveLong(template.interval)
                    || !isPositiveInt(template.connectTimeout)
                    || !isPositiveInt(template.readTimeout)) {
                throw new IllegalArgumentException("HTTP 模板不符合当前格式: " + name);
            }
        });
    }

    private boolean isPositiveLong(String value) {
        try {
            return value != null && Long.parseLong(value) > 0;
        } catch (NumberFormatException e) {
            return false;
        }
    }

    private boolean isPositiveInt(String value) {
        try {
            return value != null && Integer.parseInt(value) > 0;
        } catch (NumberFormatException e) {
            return false;
        }
    }
}

File: src/FxTools.Core/Services/HttpTemplateStore.cs
using FxTools.Core.Infrastructure;

namespace FxTools.Core.Services;

public sealed class HttpTemplateStore : IDisposable
{
    private static readonly HashSet<string> SupportedMethods =
        new(["GET", "POST", "PUT", "DELETE", "PATCH", "HEAD", "OPTIONS"], StringComparer.OrdinalIgnoreCase);
    private readonly AtomicJsonStore store;
    private readonly string path;

    public HttpTemplateStore(string? path = null)
    {
        this.path = path ?? AppDataPaths.DataFile("http_templates.json");
        store = new(this.path);
    }

    public async Task<Dictionary<string, HttpTemplate>> LoadAsync(CancellationToken cancellationToken = default)
    {
        Dictionary<string, HttpTemplate> result = await store.LoadAsync(
            () => new Dictionary<string, HttpTemplate>(StringComparer.OrdinalIgnoreCase),
            cancellationToken).ConfigureAwait(false);
        Validate(result);
        return new(result, StringComparer.OrdinalIgnoreCase);
    }

    public async Task SaveAsync(
        IReadOnlyDictionary<string, HttpTemplate> templates,
        CancellationToken cancellationToken = default)
    {
        Validate(templates);
        if (File.Exists(path))
        {
            _ = await LoadAsync(cancellationToken).ConfigureAwait(false);
        }
        await store.SaveAsync(templates, cancellationToken).ConfigureAwait(false);
    }

    private static void Validate(IReadOnlyDictionary<string, HttpTemplate> templates)
    {
        foreach ((string name, HttpTemplate template) in templates)
        {
            if (string.IsNullOrWhiteSpace(name) || name.Length > 100 || template is null
                || !Uri.TryCreate(template.Url, UriKind.Absolute, out Uri? uri)
                || uri.Scheme is not ("http" or "https")
                || !SupportedMethods.Contains(template.Method ?? string.Empty))
            {
                throw new InvalidDataException("HTTP 模板不符合当前格式。");
            }
        }
    }

    public void Dispose() => store.Dispose();
}

public sealed class HttpTemplate
{
    public string Url { get; set; } = string.Empty;
    public string Method { get; set; } = "GET";
    public string Params { get; set; } = string.Empty;
    public string Headers { get; set; } = string.Empty;
    public string Interval { get; set; } = "60";
    public string ConnectTimeout { get; set; } = "10";
    public string ReadTimeout { get; set; } = "30";
}
